<a href="https://colab.research.google.com/github/iarbietonilsson/sales-predictions/blob/main/DataScienceSalesPredictionsIAN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Science Sale Predictions
Ingrid Arbieto Nelson




# Project 1 - Parts 1 & 2

## Part 1

This is an overview of the steps you will need to complete for part one of the project. There are more detailed instructions for each section following the overview.

1) Create a GitHub Account

2) Set up a GitHub repository that will house your first portfolio project.

3) Add Markdown cells with your name and section headers to the notebook.

4) Add the Google Colab Notebook to your GitHub repository.

5) Submit the link to your GitHub repo!
***
## Part 2

We will continue to work off of the GitHub repository you created last week for your first portfolio project. The goal of your portfolio projects are to have items to show potential employers that demonstrate your data science skills and competencies.

The first project will be a sales prediction for food items sold at various stores. The goal of this is to help the retailer understand the properties of products and outlets that play crucial roles in increasing sales.

For Part 2, your task is to use Pandas to load in the data and start exploring and cleaning it.  As data scientists, about 80% of our time is spent on data preparation - so this is a very important step!

*   Download the data using the link on the Learning Platform.
    *   (Note: [Original data source](https://datahack.analyticsvidhya.com/contest/practice-problem-big-mart-sales-iii/)).
*   Insert code below the heading "Loading Data"
    *   Use Pandas to read the sales prediction data set into Google Colab.
    *   Use df.info() and df.head() methods to preview the first rows of your data and a summary of your DataFrame's columns.
*   Insert code below the heading "Data Cleaning"
    *   Explore the data - what do you need to do to clean this data? Clean it!
    *   Be sure to address each of the following tasks:

        1) How many rows and columns?

        2) What are the datatypes of each variable?

        3) Are there duplicates? If so, drop any duplicates.

        4) Identify missing values.

        5) Decide on how to address the missing values and do it! (This requires your judgement, so explain your choice).

        6) Confirm that there are no missing values after addressing them.

        7) Find and fix any inconsistent categories of data (example: fix cat, Cat, and cats so that they are consistent).

        8) For any numerical columns, obtain the summary statistics of each (min, max, mean).

    Here are a few resources to help you consider more advanced ways to deal with missing values in your dataset (check out the optional additional Pandas resources at the end of this Chapter for more resources on this):
    *   [Missing Data, its types, and statistical methods to deal with it.](https://medium.com/@raoufkeskes/missing-data-its-types-and-statistical-methods-to-deal-with-it-5cf8b71a443f#id_token=eyJhbGciOiJSUzI1NiIsImtpZCI6IjAyYTYxZWZkMmE0NGZjMjE1MTQ4ZDRlZmZjMzRkNmE3YjJhYzI2ZjAiLCJ0eXAiOiJKV1QifQ.eyJpc3MiOiJodHRwczovL2FjY291bnRzLmdvb2dsZS5jb20iLCJuYmYiOjE2NzE3OTQwNDcsImF1ZCI6IjIxNjI5NjAzNTgzNC1rMWs2cWUwNjBzMnRwMmEyamFtNGxqZGNtczAwc3R0Zy5hcHBzLmdvb2dsZXVzZXJjb250ZW50LmNvbSIsInN1YiI6IjExMjk4NTQyNDkxMTAxNjkzODM1NCIsImVtYWlsIjoic3dlZXRzYW1iYWdpcmxAZ21haWwuY29tIiwiZW1haWxfdmVyaWZpZWQiOnRydWUsImF6cCI6IjIxNjI5NjAzNTgzNC1rMWs2cWUwNjBzMnRwMmEyamFtNGxqZGNtczAwc3R0Zy5hcHBzLmdvb2dsZXVzZXJjb250ZW50LmNvbSIsIm5hbWUiOiJJbmdyaWQgTmVsc29uIiwicGljdHVyZSI6Imh0dHBzOi8vbGgzLmdvb2dsZXVzZXJjb250ZW50LmNvbS9hL0FFZEZUcDY3WnpyNEROZlQ3cHhjZ2xrZnZXUWs3SWx3MGszSDk2SG9WYmhqdEE9czk2LWMiLCJnaXZlbl9uYW1lIjoiSW5ncmlkIiwiZmFtaWx5X25hbWUiOiJOZWxzb24iLCJpYXQiOjE2NzE3OTQzNDcsImV4cCI6MTY3MTc5Nzk0NywianRpIjoiNTAwYWZhN2ZiN2M3MGUwMzBjNzJiMzQ2ZWNhZDFkZmEyOTJiN2ZlMiJ9.kZLlrhvJp0E2Rc1Kz_c4UjZ7De2E9D81MNlbouBxBJxglhhzU3ATfGO9ThjngyRmT6pDjF0JcYqmn8U-zEOEuy2PXWCGKPt056e0UYgvTAHqXR7bjbYNKWOV256ntYv2utaurDRT3d09SFO12Gh2Ou9pHvAvp-3YtrMHwQYdQFHuaedKfKSQDEad_38SbnFUIGYnMRfAJGBQpNEwz_11fMK2KnaKPMdFm9MVkWLh776r_t1GEvY-FYEvCcOvf_yQo0mzNdXA6HtPO_tNVVDj3BYCyvMKhHNdJXWUo9Y9EaQuGhF3DwvIVUP9dRecRIJoWpLclb8xoypfNIh6ID_iWQ)
    *   [6 Different Ways to Compensate for Missing Values In a Dataset.](https://https://towardsdatascience.com/6-different-ways-to-compensate-for-missing-values-data-imputation-with-examples-6022d9ca0779)
    *   [Dealing with missing data: Key assumptions and methods for applied analysis.](https://chrome-extension://efaidnbmnnnibpcajpcglclefindmkaj/https://www.bu.edu/sph/files/2014/05/Marina-tech-report.pdf)
    *   Feel free to use any other resources or techniques you find online that you think are appropriate/useful.




##Data Dictionary

![image](https://drive.google.com/uc?id=1kdsKuuQa6PA5zx6PcI_J7amlONsbFaXj)

In [70]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## Import Libraries

In [71]:
# import libraries
import pandas as pd
import numpy as np

## Loading Data

In [72]:
# load data
url = 'https://docs.google.com/spreadsheets/d/e/2PACX-1vSZM_GiLJAcSH0K70Pp1eG_J_-wht9ScYsPsaJ9IstEA2OuEYpCi4_lk6KBLc99MIA0V97E-Hs-Z84y/pub?gid=415579051&single=true&output=csv'
df = pd.read_csv(url)

In [73]:
# view header - 1st 5 rows
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [74]:
# view dataframe info - datatypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


## Data Cleaning

### 1 )  Size of Dataframe

Use df.shape() to determine how many rows and columns in the dataframe.
*   There are 8,523 rows and 12 columns.

In [75]:
df.shape

(8523, 12)

### 2 )  Data Types

Check data types for each variable.

In [76]:
df.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

### 3 )  Duplicates / Drop Irrelevant Data

Drop any duplicates, i.e. data entries that are exactly the same across all columns.
*   There are no duplicate data entries.

In [77]:
# Count any duplicate entries
df.duplicated().sum()

0

Drop irrevalent data, such as
*   values that are unique in every row
*   values that are identical across every row
*   an Unnamed column (duplicate index)
*   those that are irrevalent to the task at head


#### Drop Values that are unique in every row
*   Item_Identifiers are unique item IDs, so good to eliminate, as unique in every row.
*   Outlet Identifiers are also unique, so will delete as well.

In [78]:
# See how many different values for Item_Identifier
df['Item_Identifier'].value_counts()

FDW13    10
FDG33    10
NCY18     9
FDD38     9
DRE49     9
         ..
FDY43     1
FDQ60     1
FDO33     1
DRF48     1
FDC23     1
Name: Item_Identifier, Length: 1559, dtype: int64

In [79]:
# See how many different values for Outlet_Identifier
df['Outlet_Identifier'].value_counts()

OUT027    935
OUT013    932
OUT049    930
OUT046    930
OUT035    930
OUT045    929
OUT018    928
OUT017    926
OUT010    555
OUT019    528
Name: Outlet_Identifier, dtype: int64

#### Drop identical values
*   Outlet Type has nearly identical categories. There is not much of a  difference between a grocery store versus a supermarket, so seems good to eliminate.

In [80]:
df['Outlet_Type'].value_counts()

Supermarket Type1    5577
Grocery Store        1083
Supermarket Type3     935
Supermarket Type2     928
Name: Outlet_Type, dtype: int64

#### Drop irrevalent data
*   Outlet_Establishment_Year seems pretty irrelevant to sales predictions, so eliminating that.

#### Drop Irrevalent Columns

In [81]:
df.drop(columns = ['Item_Identifier', 'Outlet_Identifier', 'Outlet_Type', 'Outlet_Establishment_Year'], inplace = True)

In [82]:
# Check data after eliminating columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Item_Weight           7060 non-null   float64
 1   Item_Fat_Content      8523 non-null   object 
 2   Item_Visibility       8523 non-null   float64
 3   Item_Type             8523 non-null   object 
 4   Item_MRP              8523 non-null   float64
 5   Outlet_Size           6113 non-null   object 
 6   Outlet_Location_Type  8523 non-null   object 
 7   Item_Outlet_Sales     8523 non-null   float64
dtypes: float64(4), object(4)
memory usage: 532.8+ KB


#### Variables Remaining
*   Item Weight - I chose to keep item weight, as heavier items may cost more. It may not be, but I'm not a SME on food sales.
*   Item Fat Content - It may be interesting to compare diet food to yummy fat-full food.
*   Item Visibility - If an item is highly visable, it seems more likely that the customer would buy it.
*   Item Type - Which food item types sell more!
*   Item_MRP - Sales price would definitely be relavant to sales.
*   Outlet Size - How big an outlet is seems to relate to how much outlet could sell.
*   Outlet Location - Location! Location! Location! - big part of sales
*   Item_Outlet_Sales - Primary Sales datapoint!

## 4 )  Identify missing values.
*   Item weight is missing for 1463 values (17.17%)
*   Outlet size is missing for for 2410 values (28.28%)




In [83]:
# Tells number of nas in each column
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Item_Weight           7060 non-null   float64
 1   Item_Fat_Content      8523 non-null   object 
 2   Item_Visibility       8523 non-null   float64
 3   Item_Type             8523 non-null   object 
 4   Item_MRP              8523 non-null   float64
 5   Outlet_Size           6113 non-null   object 
 6   Outlet_Location_Type  8523 non-null   object 
 7   Item_Outlet_Sales     8523 non-null   float64
dtypes: float64(4), object(4)
memory usage: 532.8+ KB


In [84]:
# Sums number of missing values per column
df.isna().sum()

Item_Weight             1463
Item_Fat_Content           0
Item_Visibility            0
Item_Type                  0
Item_MRP                   0
Outlet_Size             2410
Outlet_Location_Type       0
Item_Outlet_Sales          0
dtype: int64

In [85]:
# show percentage of na values / total count of item
df.isna().sum() / len(df) * 100

Item_Weight             17.165317
Item_Fat_Content         0.000000
Item_Visibility          0.000000
Item_Type                0.000000
Item_MRP                 0.000000
Outlet_Size             28.276428
Outlet_Location_Type     0.000000
Item_Outlet_Sales        0.000000
dtype: float64

## 5 ) How to Address Missing Values

For missing values, I could
1.   **Drop the column**
     *  ( _good rule of thumb < 5%_ )
2.   **Impute the missing value**
     *  many statitical methods
     *  simpler one is to use median (_integers_), mean (_float_), mode (_category_)

### Item Weight

*   Item Weight is missing for 17.17% of values, so not minimal number of missing values.
*   Chosen method - **Impute the missing value** with mean for _float_ value
*   This does introduce errors into the dataset.

In [86]:
df['Item_Weight'].fillna(value = df['Item_Weight'].mean(), inplace = True)


###Outlet Size
*   Outlet Size has 28.28% missing values, not minimal.
*   Chosen method - - **Impute the missing value** with mode for _category_ value
*   This does introduce errors into the dataset.

In [87]:
# Medium is the most common value
df['Outlet_Size'].value_counts()

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64

In [88]:
# fill in with mode - most common value
df['Outlet_Size'].fillna(value = 'Medium', inplace = True)

## 6 ) Final check for Missing Values

Check the info on the dataframe to ensure there are no longer any missing values!!

In [89]:
# Check that all values filled in
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Item_Weight           8523 non-null   float64
 1   Item_Fat_Content      8523 non-null   object 
 2   Item_Visibility       8523 non-null   float64
 3   Item_Type             8523 non-null   object 
 4   Item_MRP              8523 non-null   float64
 5   Outlet_Size           8523 non-null   object 
 6   Outlet_Location_Type  8523 non-null   object 
 7   Item_Outlet_Sales     8523 non-null   float64
dtypes: float64(4), object(4)
memory usage: 532.8+ KB


## 7 ) Find and fix any inconsistent categories of data

From data info above, floats look consistent:
*   Item Weight
*   Item Visability
*   Item_MRP
*   Item_Outlet_Sales

Next, check categorical values for inconsistent data.
*   Only Item Fat Content has inconsistent values.

In [90]:
# values for item fat content
df['Item_Fat_Content'].value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

In [91]:
# replace inconsistent values
df.replace(['LF','low fat'], ['Low Fat', 'Low Fat'], inplace = True)
df.replace(['reg'], ['Regular'], inplace = True)

In [92]:
# check after replacing for consistent values
df['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

In [93]:
# values for item type
df['Item_Type'].value_counts()

Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64

In [94]:
#  values for outlet size
df['Outlet_Size'].value_counts()

Medium    5203
Small     2388
High       932
Name: Outlet_Size, dtype: int64

In [95]:
# values for outlet location type
df['Outlet_Location_Type'].value_counts()

Tier 3    3350
Tier 2    2785
Tier 1    2388
Name: Outlet_Location_Type, dtype: int64

## 8 ) For any numerical columns, obtain the summary statistics of each (min, max, mean).

In [96]:
# summary statistics for numerical values
df.describe().round(2)

Item_Weight  Item_Visibility  Item_MRP  Item_Outlet_Sales
count      8523.00          8523.00   8523.00            8523.00
mean         12.86             0.07    140.99            2181.29
std           4.23             0.05     62.28            1706.50
min           4.56             0.00     31.29              33.29
25%           9.31             0.03     93.83             834.25
50%          12.86             0.05    143.01            1794.33
75%          16.00             0.09    185.64            3101.30
max          21.35             0.33    266.89           13086.96

In [97]:
# another way to show min, max, & mean
summary_stats = df.describe()
summary_stats.loc[['min','max','mean']]

Item_Weight  Item_Visibility    Item_MRP  Item_Outlet_Sales
min      4.555000         0.000000   31.290000          33.290000
max     21.350000         0.328391  266.888400       13086.964800
mean    12.857645         0.066132  140.992782        2181.288914

## Exploratory Visuals

## Explanatory Visuals